### ISSUE TITLE
Add AOT relocation support for Immediate instructions on Z 

### ISSUE THREAD

**Issue by r30shah** [eclipse-openj9/openj9#4391]
As per discussion in https://github.com/eclipse/openj9/pull/4310, Currently on Z, whenever we are loading a constant address, we use snippet in case we are generating relocatable code. We should add relocation support in instructions which can be used to load constant address (J9Class or J9Method) from immediate field so we can have better AOT compiled body.

**Comment by dchopra001** [comment]
As part of this work we should add AOT and remote compilation support for PC-relative instructions as well.

Right now, the story for relative instructions on Z is as follows:

1. If we are using a `BRASL` to branch to a method, then we can relocate that method call.
2. If we are trying to reach an address within the JITT'ed method, we don't need a relocation as the relocation as the offset remains the same across runs.
3. If we are trying to load an address that is outside of the JITT'ed method body, then we cannot relocate that instruction. We should be able to use a similar approach to the one in '1.` above for `BRASL` instructions. To guard against generating these instructions for remote compilations, we use the `cg->canUseRelativeLongInstructions()` query at the moment.

**Comment by dchopra001** [comment]
I have been looking into this issue recently. I was initially investigating using relative instructions (ex. `LARL`) to materialize address values during runtime. This is not possible for any AOT compile because we don't know where the code cache will be at compile time. If the address we are trying to get to is not reachable using an `LARL`, then a relocation won't be possible. Since we can't make this decision at compile time, relative instructions can't be used for AOT compiles. An alternative solution is to use two load immediate instructions to materialize the address. Something similar is done on Power as they didn't have support for relative instructions for this type of use until recently. I suspect we didn't do this already on X/Z because support for relative instructions was available, and it probably wasn't worth the performance gain to solve this problem with added complexity. We should still be able to use LARL for JITServer compiles. I'll have a PR for that shortly.

**Comment by r30shah** [comment]
I agree with the Branch instructions, we do not know where target will be in code cache though. Is it also true in case we are loading constant such as class pointer, I thought on x86 we were able to use the MoveRegImm instructions for constant address load [1] , as we do add relocation for them [2]. I remember, I opened up this issue to address load address constants as I observed x86 would generate load immediate instructions. I am not sure if I have a log file handy but it should not be difficult to verify. 

On Z for loading such constant we call `generateRegLitRefInstruction` [3] which allocates snippet and later generates `LG` or `LGRL` to load the constant. 


[1]. https://github.com/eclipse/omr/blob/e2fac34fcb3571cdf5351693a04fc2852b18f46f/compiler/x/codegen/OMRTreeEvaluator.cpp#L177
[2]. https://github.com/eclipse/omr/blob/e2fac34fcb3571cdf5351693a04fc2852b18f46f/compiler/x/codegen/X86BinaryEncoding.cpp#L1653

**Comment by dchopra001** [comment]
> I remember, I opened up this issue to address load address constants as I observed x86 would generate load immediate instructions

I believe on x86 we have MOV instructions that can move a 64-bit value into a register. As we already know the same is not true for Z. On Z we would have to use two instructions to do the same. This could still be beneficial.

> Just to verify I am talking about the class pointer constants which are guaranteed to be relative addressable in compressed refs.
I suppose if we limit to compressed refs, then we may be able to use relative instructions all the time. But if we do this then special casing for compressed refs could be another added complexity that we may not need.

**Comment by r30shah** [comment]
> I suppose if we limit to compressed refs, then we may be able to use relative instructions all the time. But if we do this then special casing for compressed refs could be another added complexity that we may not need.

IMO there is still a benefit from using Load Relative on J9Method pointers and J9Class pointers as it saves a memory load and 8 byte of code cache used for the snippet. Not sure how many places this changes can be applied and benefit would be that visible. 
Just FYI, for `BRASL` we do add relocation at binary encoding time [1], and for AOT compiles we use them for helper calls. 

[1]. https://github.com/eclipse/omr/blob/bb24584261c08c85ab0e8fc4352d03c38b6375b7/compiler/z/codegen/S390Instruction.cpp#L2264-L2274

**Comment by dsouzai** [comment]
I believe the reason we can use `BRASL` for helper calls is because we will create a trampoline if needed in the relocation run; @dchopra001 noticed this happens in [1]. Trampolines are allocated in a separate space in the code cache so this is possible; however, if the address to be loaded can't fit into the immediate of `LARL`, then there's no way we can allocate a data snippet at relo time. We can't assume that the PC relative offset will always fit in `LARL` because in different JVM instances, where the codecache, jit dll, and java heap end up are completely arbitrary with ASLR.


[1] https://github.com/eclipse/openj9/blob/bb626b213f343b38bb3a610b5cdf86a50f8f0174/runtime/compiler/runtime/RelocationRecord.cpp#L600
